File: incre-tests/autolifter/single-pass/longest00s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

is_even = \a: Int. == a (* 2 (/ a 2));

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

longest00s = fix (
  \f: Int -> List -> Int.
  \len: Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let len = if == h 0 then + len 1 else 0 in
    if is_even len then max len (f len t)
    else f len t
  end
) 0;

main = single_pass longest00s;

File: ml-benchmark/autolifter/single-pass/longest00s.ml
type l = Cons of int * l | Nil of unit

let is_even a = a = 2 * (a / 2)

let single_pass v xs =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  v (run xs)

let inf = 100

let max a b = if a < b then b else a

let longest00s =
  let rec f len xs =
    match xs with
    | Nil _ -> 0
    | Cons (h, t) ->
      let len = if h = 0 then len + 1 else 0 in
      if is_even len then max len (f len t) else f len t
  in
  fun xs -> f 0 xs

let program = single_pass longest00s